### PR TITLE
feat: some users don't like dark color theme - #3

### PR DESCRIPTION
**Pull request by power-f-GOD**
This PR features the dark-light mode theme toggle.

Summary of the changes made in this PR:

1. [feat: Add `SVGIcon` component and theme toggle button](https://github.com/holdex/holdex-venture-studio/commit/da746f55205537597fcb4ed223ebc77f24dc31c8) 

    - Augment `tailwind.config#theme.screens` with breakpoints that match actual Figma and adhere to the Mobile-first RWD principle;
    - Remove repetitions and redundant nodes and classNames in nav UI implementation in root `layout.pug`;
    - Clean up and modify nav source implementation.

2. [feat: Add theme toggle logic](https://github.com/holdex/holdex-venture-studio/commit/bc65ce25d107c37c50e1342e8adcb64e56fe1da3) 

    - Create `colorScheme` mixin in `theme.scss`;
    - Remove `data-theme` attributes on `.app.container` in `error.pug` and `layout.pug`.

3. [feat: Complete dark-light theme toggle feature](https://github.com/holdex/holdex-venture-studio/commit/4932679f75ddc9fe6c7ec6bb23e71e518bffbe0b) 

    - Add `darkMode` config prop in `tailwind.config`;
    - Update button shadow class alpha values;
    - Update all colors (and color classNames) to match that provided in the Figma (for light theme);
    - Add Holdex logo to `SVGIcon` for dark/light variant sake;
    - Finish solving the problem.

As always, let me know if there are any updates you want me to make.🙇🏽‍♂️

Gracias.🙌🏼

### CODE REVIEW

File: tailwind.config.cjs [hunk 1]
```diff
@@ -13,6 +14,13 @@ module.exports = {
 			underbase: '11px',
 		},
 		screens: {
+			// 'rwd' means "Responsive Web Design". And these breakpoints follow/match the breakpoints in the actual Figma while adhering to the Mobile-first RWD principle.
+			'rwd-xs': '375px',
+			'rwd-sm': '480px',
+			'rwd-md': '768px',
+			'rwd-lg': '992px',
+			'rwd-xl': '1200px',
+			'rwd-2xl': '1440px',
```
Comment (georgeciubotaru) [inline tailwind.config.cjs:23]: @power-f-GOD Don't think we need these as we already have our own breakpoints

File: src/lib/components/Icons/SVGIcon.svelte [hunk 2]
```diff
@@ -0,0 +1,78 @@
+<script lang="ts">
+	/** internal deps */
```
Comment (georgeciubotaru) [inline src/lib/components/Icons/SVGIcon.svelte:2]: @power-f-GOD We have a button, type `solo`.
Also we are using `hero Icons`, imported via `$lib/components/Icons`

File: tailwind.config.cjs (hunk 1 above)
Comment (power-f-GOD) [reply tailwind.config.cjs:23]: Ok. Interesting that you didn't highlight the comment I dropped there just before these, @georgeciubotaru .😅

Here you go...

```js
  ...
  screens: {
    // 'rwd' means "Responsive Web Design". And these breakpoints follow/match the breakpoints in the actual Figma while adhering to the Mobile-first RWD principle.
    'rwd-xs': '375px',
    'rwd-sm': '480px',
    'rwd-md': '768px',
    'rwd-lg': '992px',
    'rwd-xl': '1200px',
    'rwd-2xl': '1440px',
    ...
```

So, for this, we might have to "disagree" to agree.🥲

Currently, the breakpoints that are in the `tailwind.config` don't follow, strictly, the Mobile-first RWD principle, and is not well structured to follow the same. From (and with) my experience with building UI's and following this principle, I couldn't work with those breakpoints. And of course, I couldn't modify them as doing so would break existing UI/design, so I thought to add the breakpoints above.

However, what I can do is probably change the name from `rwd-xs` to `xs-up` etc. if you're not comfortable with `rwd`, since we have similar naming (`smDown`) currently.

Cc: @zolotokrylin

Comment (power-f-GOD) [reply tailwind.config.cjs:23]: Ok. Interesting that you didn't highlight the comment I dropped there just before these, @georgeciubotaru .😅

Here you go...

```js
  ...
  screens: {
    // 'rwd' means "Responsive Web Design". And these breakpoints follow/match the breakpoints in the actual Figma while adhering to the Mobile-first RWD principle.
    'rwd-xs': '375px',
    'rwd-sm': '480px',
    'rwd-md': '768px',
    'rwd-lg': '992px',
    'rwd-xl': '1200px',
    'rwd-2xl': '1440px',
    ...
```

So, for this, we might have to "disagree" to agree.🥲

Currently, the breakpoints that are in the `tailwind.config` don't follow, strictly, the Mobile-first RWD principle, and is not well structured to follow the same. From (and with) my experience with building UI's and following this principle, I couldn't work with those breakpoints. And of course, I couldn't modify them as doing so would break existing UI/design, so I thought to add the breakpoints above.

However, what I can do is probably change the name from `rwd-xs` to `xs-up` etc. if you're not comfortable with `rwd`, since we have similar naming (`smDown`) currently.

Cc: @zolotokrylin

File: src/routes/(pages)/+layout.svelte [hunk 3]
```diff
@@ -88,8 +97,15 @@
...
 
+	/** lifecycles */
+	onMount(() => {
+		themeIconName = window.matchMedia('(prefers-color-scheme: dark)').matches ? 'sun' : 'moon';
+	});
+
+	/** react-ibles */
 	$: path = $page.url.pathname;
 	$: form = $page.form;
+	$: if (isBrowser) document.body.dataset.theme = themeIconName === 'moon' ? 'light' : 'dark';
```
[the remark this answers is not in the document]
Comment (power-f-GOD) [reply src/routes/(pages)/+layout.svelte:108]: Actually, no. The idea is for it to run every time `themeIconName` changes. Putting it in the `onMount` will make the theme toggle button useless.🥲

You grab?

File: tailwind.config.cjs (hunk 1 above)
Comment (georgeciubotaru) [reply tailwind.config.cjs:23]: @power-f-GOD We have our own breakpoints.
The design in Figma is made for these breakpoints, so we shouldn't add additional breakpoints without a special need, that should be reported (so the Designer can be aware).

The current breakpoints cover all the needs, so just use them

File: src/routes/(pages)/+layout.svelte (hunk 3 above)
[the remark this answers is not in the document]
Comment (georgeciubotaru) [reply src/routes/(pages)/+layout.svelte:108]: Got you

**Review by georgeciubotaru — CHANGES_REQUESTED** [review]
left comments

File: src/lib/components/Icons/SVGIcon.svelte (hunk 2 above)
Comment (power-f-GOD) [reply src/lib/components/Icons/SVGIcon.svelte:2]: Oh. Yeah, thanks for the URL. Those are the icons I actually used which I got/copied from the Figma design, and they are the same in the URL you provided.🤲🏼

File: tailwind.config.cjs (hunk 1 above)
Comment (power-f-GOD) [reply tailwind.config.cjs:23]: Also, I forgot to mention that with these breakpoints I added, I was able to remove a significant amount of redundant class names and elements/nodes, as you must have seen in my PR.🙇🏽‍♂️

Comment (power-f-GOD) [reply tailwind.config.cjs:23]: @georgeciubotaru Yes, correct. We should be (and remain) open-minded.✔️ Also, we should see reasons as well.🤲🏼

I'll answer your questions by paragraph...

First,

I don't think it's a nice thing to water down (or reduce) what I did in this task to compare it with "only a solo icon button...and color changes." If it were so, I guess it should have been solved long before I hopped on it, and even if it were solved long before it still would not have been nice to reduce the effort put in.

Any designs done on Figma target multiple screen sizes and not one screen in particular, hence it was necessary to complete the task based on the screen sizes.

> this scope doesn't imply any big changes or changes to other aspects of the project

Ok. I'm not sure what other aspects you're talking about, but my focus in this task was mainly on the problem and on (surrounding) code that affects it (in the case of the button). You saw it in my PR. And if it's for the new breakpoints including, I gave my reasons already above which was necessary for the completion of my task.

Second,

I didn't say there was anything wrong with/on the (Figma) Design perspective, but there was on the code perspective (the code/config design/implementation of the Design breakpoints in the `tailwind.config` which I have also stated/explained in my previous comments). 
(And BTW, the only thing I said/did was to give a suggestion on the Design to better the UX of the user. I didn't say something was wrong with the Design.)📝

Third,

> There is no right way to do things

Well, you may (or may not) be right. Notwithstanding, if there's no "right way" of doing things, there's surely a "standard way" of doing things. Yeah?

And your assertion of being a perfectionist...hmm...Ok. Not sure what to say about that😅, but I believe all of us in, some way, are perfectionists and strive to do things perfectly. An example is that we strive/aim to deploy bug-free code, that's perfectionism in a sense.😃 But that's by the way.😌
Notwithstanding, if we can do things the "right"/"standard"/"perfect" way the first time (without any "costs"), I think we shouldn't hold back, and this doesn't stop us from solving problems (or being problem solvers).🤲🏼

Fourth,

Yeah, sure. Of course. I wouldn't add these breakpoints if I didn't need them or if (the lack of them) didn't affect my task. You can take a look at my PR where I implemented them and let me know if it was unnecessary or not needed for the task. Again, I explained the reason(s) above.

Thanks for the comments/review though.🙇🏽‍♂️ The back and forths are sometimes necessary to come to a compromise.🙂

**Comment by georgeciubotaru** [comment]
@predrag-us Please check the UI, preview URL is available

**Comment by zolotokrylin** [comment]
Problem: browser doesn't remember my color scheme preference

To be resolved in a separate PR (not in this one).

**Comment by power-f-GOD** [comment]
> Problem: browser doesn't remember my color scheme preference
> 
> To be resolved in a separate PR (not in this one).

Yeah, I was going to mention this in my PR summary, but I chose to wait until one of you mentions it, and now you have!😃

So, since you've mentioned that. I'll take it up in another Problem.✔️

BTW, FWIW, the reason I implemented it the current way is that I thought it would be nice/good to always default to the user's OS's theme mode on initial load of the page, and if they want otherwise, they can always toggle the button.
Basically, I thought the user's OS theme mode to be my "`localStorage`", in a sense.🙃

But, in the new problem solution, I'll make it so that when they interact with the theme button, I'll save their chosen theme in `localStorage`, and use that on page reload, and if they don't, always default to their OS's theme mode/setting.📝

What do you reckon, @zolotokrylin ? Good?

File: tailwind.config.cjs (hunk 1 above)
Comment (georgeciubotaru) [reply tailwind.config.cjs]: @power-f-GOD 

Sorry if you're getting this feeling but I was not trying to reduce your efforts in any way. I hope you don't mid our little back & forth. Previously I just resumed the Problem solution in a single sentence. Of course I can see your PR and notice that you added the handler for theme switching, you updated the classes to support theme colors and you added an icon that we discussed about. You are going in the right direction. 👍

What I was trying to say, is that we shouldn't deviate from the Problem/Solution statement. If a Problem can be solved separately, we should aim to not mix it within the current Problem statement. Unless it's coupled with the current problem, then take one of  3 decisions: 
1) if it blocks and enlarges the scope - fix it first separately and then come back to the original problem
2) if it block but doesn't enlarge the scode - fix it inside the current scope as quick as possible so other members are not blocked by it.
3) If it doesn't block - create a separate problem statement.

About design,

I didn't said either that something is wrong with it. I just wanted to point that the process starts with Design first and not vice-versa. This means, if a visual problem was encountered in the code, most likely it's also in the design and needs to be solved there in the first place. So, if you noticed that we are missing a breakpoint, this needs to be addressed in the design first in a separate problem. I personally didn't encounter this problem and I believe our design team has put thought into their decisions too. There is nothing bad about giving suggestion or feedback to design. We are all transparent inside the team and feedback is never personal related, it's always about solving the problem. 💪

About "standard way",

It all depends. When it comes to our industry, everything depends. You will never find the truth. I'm not ideological and I treat technology as tradeoffs. What I mean is that there might exists standards like W3C standards or Google invented standards or Facebook invented standards who obviously put some thought in it. But I try to remain open and flexible here without any harm to the product and code. I encountered times when things can be more simple outside the standard, or that the standards are overkilling. I also do my best to implement things correctly, give the right advice, deploy bug-free code and test things before launch. It's good to know the standards but it's also good to know when is the best time to apply them. 

I like your approach and I will support your attitude of doings things the right way from the first time without costs. I also like that you have a strong point of view, keep it up. One of our team values on About page says "do the right things. because sometimes doings things right can be in vain" ✌️

--- 

I believe the problem could be solved without the breakpoints. I will leave you comments in the necessary places.

**Comment by georgeciubotaru** [comment]
@power-f-GOD Theme should be stored in cookie

**Comment by power-f-GOD** [comment]
> @power-f-GOD Theme should be stored in cookies

Ok. That works too. ✅🙌🏼

File: tailwind.config.cjs (hunk 1 above)
Comment (zolotokrylin) [reply tailwind.config.cjs]: @georgeciubotaru thanks for explaining our principles.
@power-f-GOD let me summarise (and rephrase) in a shorter version what George just said:
1. When we solve the problem, we MUST (not SHOULD, but MUST) solve exactly that problem. When we solve the current problem, we might face or notice other problems along the way. This is a great opportunity to report new problems.
3. New problems MUST be reported in a separate issue with a reference to the related issue (if any). The best way to do it is by posting comments on the current issue and then creating a new issue ([instruction how to do it](It is more simple and keeps things in context.).
4. When you start working on the issue before you start coding you SHOULD review the design and if you find errors in the design, better to report it back to the designers (you can ping Predrag, Mark, George, or me).
5. We are open to new suggestions regarding standards (we call them "principles"), but this suggestion SHOULD be applied by proposing a problem (like you did with indentation #375). Unless that proposal went from BACKLOG to DONE it is not our principles, and thus, current principles are the rules. This way allows us to move swiftly and not get stuck in discussions about new approaches ("there are as many opinions as there are experts").

And good stuff, I love that you have many proposals! ❤️  The next step is to channel your new ideas and suggestion into separate independent github issues (problem statements) which can be reviewed independently!

**Comment by predrag-us** [comment]
Hey @power-f-GOD, great work! Love seeing our website in the white coat! ✨

Here're a few things I've noticed: 

- Please check the size of the `themeSwitch` button.
- After switching to `light` there's an unnecessary border around the `themeSwitch` icon which is not visible when switching to `dark` mode.
- Looks like you've applied a `.svg` animation on the icons while switching, although I'm not sure. In case I'm right, please remove the animation. 
- Please be aware of shadows. We do have shadows on `dark` but should we have them on `light`?
- Please make sure to double-check the colors. Blockquote is one example where colors don't match the design. There are more examples, I'll let you find them.
- Be aware of the hover and active states. Link is one example where the hover state is missing. Button shadows on hover are incorrect. There are more examples, I'll let you find them, again.

The bottom line ⏤ applying a `light` theme isn't simply switching colors, you have to be aware of other properties as well. Always double-check your work and make sure it matches the design. 

Hope you'll find this helpful. Thank you!

File: tailwind.config.cjs (hunk 1 above)
Comment (power-f-GOD) [reply tailwind.config.cjs]: @zolotokrylin , just saw your comment.😌 Yeah, very much understood and noted.📝 Thank you!🙌🏼💕

**Comment by power-f-GOD** [comment]
> Hey @power-f-GOD, great work! Love seeing our website in the white coat! ✨
> 
> Here're a few things I've noticed:
> 
> * Please check the size of the `themeSwitch` button.
...

Ok. That's a lot.😅 Talk about perfectionism.🙃 I like it though.😊

All right. Noted. Will have a look again.📝

BTW, @predrag-us , if I could get an interaction video (or flow) in the Figma, it would be helpful in letting me know what these buttons, colors, shadows etc., basically states, should look like when they are interacted with. The current static version of the design keeps me in a meta-guess-verse.🥲

**Comment by power-f-GOD** [comment]
@predrag-us ,

> - [x] Please check the size of the themeSwitch button.

> - [ ] After switching to light there's an unnecessary border around the themeSwitch icon which is not visible when switching to dark mode.

Ok. I'm not sure I understand this🤔, but is the border different from what is in this screenshot?

Screenshot 2023-06-15 at 3 23 26 AM

> - [x] Looks like you've applied a .svg animation on the icons while switching, although I'm not sure. In case I'm right, please remove the animation.

> - [x] Please be aware of shadows. We do have shadows on dark but should we have them on light?

> - [ ] Please make sure to double-check the colors. Blockquote is one example where colors don't match the design. There are more examples, I'll let you find them.

Ok. I couldn't find this one. Can you show me the examples (by screenshots or a video)?

> - [x] Be aware of the hover and active states. Link is one example where the hover state is missing. Button shadows on hover are incorrect. There are more examples, I'll let you find them, again.

Ok. This was missing (in some places) before I hopped on solving the context problem, but I have found and fixed it. And it was caused by some missing spacing between two class names which caused the hover style to not be applied on hover...

Before:
```js
  ...
  classes += 'text-paragraph-l';
  ...
```

Now:
```js
  ...
  classes += ' text-paragraph-l';
  ...
```

**Comment by predrag-us** [comment]
@power-f-GOD 

> The current static version of the design keeps me in a meta-guess-verse.🥲

That shouldn't be the case. Each component has properties, including the `state`. Here's an example of the button:

CleanShot 2023-06-15 at 09 18 23@2x

---

> After switching to `light` there's an unnecessary border around the `themeSwitch` icon which is not visible when switching to `dark` mode.

Please have a look here: https://share.cleanshot.com/gWxNH1XT

---

> Please make sure to double-check the colors. Blockquote is one example where colors don't match the design. There are more examples, I'll let you find them.

Implementation: 

CleanShot 2023-06-15 at 09 24 09@2x

Design: 

CleanShot 2023-06-15 at 09 24 31@2x

---

Hope this helps. Thank you!

**Comment by power-f-GOD** [comment]
@predrag-us , Ok. About the border on the button, I don't see what you're referring to. The Figma has a (blue) border for its hover/focus state...what am I doing differently, or which exactly are you referring to?🤔

Screenshot 2023-06-15 at 4 46 11 PM

For the blockquote background, yeah, will have a look and fix in a moment.🤲🏼

**Comment by predrag-us** [comment]
@power-f-GOD  Yes, it's good that we have a border on hover, but please notice that the border stays after I click the button, while that's not the case when switching to `dark` theme, which is current.

Regarding the blockquote component background, it's not only blockquote, I just gave you one example where this problem occurs. It may be that the issue is actually the body background, as previously mentioned.

Hope this helps. Thank you!

**Comment by zolotokrylin** [comment]
Guys, remember: invest time what makes the difference right now. 
We don't want to spend time on things which will not drive us more business. If you wish, create new problem issue and I will assign it priority. Let's merge this PR today max(!).

**Comment by power-f-GOD** [comment]
I think I've fixed all fixables/requests and PR can be merged.✔️

Moving on to another problem.🤲🏼

**Comment by zolotokrylin** [comment]
Well done guys!
